File: jaxrate/presets.py
"""Pre-built grammars for common phylogenetic annotation tasks."""

import jax.numpy as jnp
import math

from .types import Grammar, Nonterminal, Rule, EmissionGroup
from .grammar import GrammarBuilder


def pfold_grammar():
    """PFOLD-like RNA structure SCFG.

    Nonterminals:
        S — start/structure (can be stem or loop)
        L — stem (paired emissions)
        F — loop/unpaired region

    Rules:
        S → L S        (stem followed by more structure)
        S → F          (loop region)
        L → e₁ S e₂   (paired emission flanking inner structure)
        F → e F        (emit unpaired, continue loop)
        F → e          (emit unpaired, end)

    Models:
        0 — unpaired emission model
        1 — paired emission model

    Returns:
        Grammar
    """
    gb = GrammarBuilder()

    S = gb.add_nonterminal('S')
    L = gb.add_nonterminal('L')
    F = gb.add_nonterminal('F')

    # S → L S  (stem then more)
    gb.add_rule(S, rhs=[L, S], log_weight=jnp.log(0.4))

    # S → F  (just a loop)
    gb.add_rule(S, rhs=[F], log_weight=jnp.log(0.6))

    # L → e₁ S e₂  (paired emission flanking S)
    gb.add_rule(L, rhs=[S],
                emissions=[EmissionGroup(n_positions=2, model_index=1)],
                log_weight=0.0)

    # F → e F  (emit unpaired, continue)
    gb.add_rule(F, rhs=[F],
                emissions=[EmissionGroup(n_positions=1, model_index=0)],
                log_weight=jnp.log(0.6))

    # F → e  (emit unpaired, end)
    gb.add_rule(F, rhs=[],
                emissions=[EmissionGroup(n_positions=1, model_index=0)],
                log_weight=jnp.log(0.4))

    return gb.build(start=S, n_models=2)


def gene_finder_grammar():
    """Gene-finding HMM.

    States:
        IG — intergenic (single nucleotide emission)
        E1, E2, E3 — exon codon positions 1, 2, 3
        I — intron (single nucleotide emission)

    Transitions model a simplified gene structure:
        IG → IG | E1
        E1 → E2
        E2 → E3
        E3 → E1 | I | IG
        I → I | E1

    Models:
        0 — intergenic model (single nucleotide)
        1 — exon model (single nucleotide, codon-position-specific)
        2 — intron model (single nucleotide)

    For codon-level phylogenetic models, the caller should pre-compute
    terminal weights using kmer_tokenize with stride=1 for codon windows.

    Returns:
        Grammar
    """
    gb = GrammarBuilder()

    IG = gb.add_nonterminal('IG')
    E1 = gb.add_nonterminal('E1')
    E2 = gb.add_nonterminal('E2')
    E3 = gb.add_nonterminal('E3')
    I = gb.add_nonterminal('I')

    p_gene = 0.01      # probability of starting a gene
    p_stop = 0.01      # probability of stopping a gene
    p_intron = 0.02    # probability of entering intron
    p_exon_resume = 0.05  # probability of leaving intron

    # IG → emit IG
    gb.add_rule(IG, rhs=[IG],
                emissions=[EmissionGroup(1, 0)],
                log_weight=math.log(1 - p_gene))
    # IG → emit E1
    gb.add_rule(IG, rhs=[E1],
                emissions=[EmissionGroup(1, 0)],
                log_weight=math.log(p_gene))
    # IG → emit (terminal)
    gb.add_rule(IG, rhs=[],
                emissions=[EmissionGroup(1, 0)],
                log_weight=math.log(1e-10))  # very small term prob

    # E1 → emit E2
    gb.add_rule(E1, rhs=[E2],
                emissions=[EmissionGroup(1, 1)],
                log_weight=0.0)

    # E2 → emit E3
    gb.add_rule(E2, rhs=[E3],
                emissions=[EmissionGroup(1, 1)],
                log_weight=0.0)

    # E3 → emit E1 (continue codon)
    gb.add_rule(E3, rhs=[E1],
                emissions=[EmissionGroup(1, 1)],
                log_weight=math.log(1 - p_stop - p_intron))
    # E3 → emit I (enter intron)
    gb.add_rule(E3, rhs=[I],
                emissions=[EmissionGroup(1, 1)],
                log_weight=math.log(p_intron))
    # E3 → emit IG (stop gene)
    gb.add_rule(E3, rhs=[IG],
                emissions=[EmissionGroup(1, 1)],
                log_weight=math.log(p_stop))
    # E3 → emit (terminal)
    gb.add_rule(E3, rhs=[],
                emissions=[EmissionGroup(1, 1)],
                log_weight=math.log(1e-10))

    # I → emit I
    gb.add_rule(I, rhs=[I],
                emissions=[EmissionGroup(1, 2)],
                log_weight=math.log(1 - p_exon_resume))
    # I → emit E1 (resume exon)
    gb.add_rule(I, rhs=[E1],
                emissions=[EmissionGroup(1, 2)],
                log_weight=math.log(p_exon_resume))

    return gb.build(start=IG, n_models=3)


def pseudoknot_grammar(max_span=None):
    """Pseudoknot-capable MCFG with fan-out 2.

    Nonterminals:
        S — start (fan-out 1): generates full sequence
        PK — pseudoknot (fan-out 2): generates two crossing stem-loop regions
        L — stem (fan-out 1): generates a stem-loop

    Rules:
        S → L S       (stem-loop followed by more structure)
        S → PK S      (pseudoknot followed by more structure)
        S → e         (single unpaired emission)
        L → e₁ S e₂   (paired emission flanking inner structure)
        PK(x, y) → L(x) L(y)  (two crossing stems)

    Models:
        0 — unpaired emission model
        1 — paired emission model

    Args:
        max_span: Maximum span per component for the PK nonterminal.
                  None means unlimited. Constraining this reduces complexity
                  from O(C⁶K³) to O(C²L²K³ + C³K³).

    Returns:
        Grammar
    """
    gb = GrammarBuilder()

    S = gb.add_nonterminal('S', fan_out=1)
    PK = gb.add_nonterminal('PK', fan_out=2, max_span=max_span)
    L = gb.add_nonterminal('L', fan_out=1)

    # S → L S  (stem then more structure)
    gb.add_rule(S, rhs=[L, S], log_weight=jnp.log(0.2))

    # S → e S  (unpaired then more structure)
    gb.add_rule(S, rhs=[S],
                emissions=[EmissionGroup(1, 0)],
                log_weight=jnp.log(0.3))

    # S → PK  (pseudoknot region, concatenates PK's two components)
    gb.add_rule(S, rhs=[PK], log_weight=jnp.log(0.1))

    # S → e (terminal unpaired)
    gb.add_rule(S, rhs=[],
                emissions=[EmissionGroup(1, 0)],
                log_weight=jnp.log(0.4))

    # L → e₁ S e₂ (paired emission flanking inner structure)
    gb.add_rule(L, rhs=[S],
                emissions=[EmissionGroup(2, 1)],
                log_weight=0.0)

    # PK(ax, by) → PK(x, y) [pair(a,b)]  (extend stem 1: left-left)
    # Pairs left of component 1 with left of component 2
    gb.add_rule(PK, rhs=[PK],
                emissions=[EmissionGroup(2, 1)],
                composition=('ll',),
                log_weight=jnp.log(0.4))

    # PK(xa, yb) → PK(x, y) [pair(a,b)]  (extend stem 2: right-right)
    # Pairs right of component 1 with right of component 2
    gb.add_rule(PK, rhs=[PK],
                emissions=[EmissionGroup(2, 1)],
                composition=('rr',),
                log_weight=jnp.log(0.4))

    # PK(x, y) → S(x) S(y)  (base: each component is independent structure)
    gb.add_rule(PK, rhs=[S, S], log_weight=jnp.log(0.2))

    return gb.build(start=S, n_models=2)

File: web/src/presets.ts
/**
 * Pre-built grammars: pfold, gene_finder, pseudoknot.
 * Mirrors jaxrate/presets.py.
 */

import type { Grammar, EmissionGroup } from './types.js';
import { GrammarBuilder } from './grammar.js';

const emit1 = (model: number): EmissionGroup => ({ nPositions: 1, modelIndex: model });
const emit2 = (model: number): EmissionGroup => ({ nPositions: 2, modelIndex: model });

/**
 * RNA structure prediction SCFG (pfold-like).
 * Models: 0=unpaired, 1=paired.
 *
 * S → L S | F
 * L → e₁ S e₂ (paired)
 * F → e F | e (unpaired)
 */
export function pfoldGrammar(): Grammar {
  const gb = new GrammarBuilder();
  const S = gb.addNonterminal('S');
  const L = gb.addNonterminal('L');
  const F = gb.addNonterminal('F');
  const Fstar = gb.addNonterminal('F*');

  // S → L S (bifurcation into stem + continue)
  gb.addRule(S, [L, S], [], Math.log(0.4));
  // S → F (single-stranded)
  gb.addRule(S, [F], [], Math.log(0.5));
  // S → ε
  gb.addRule(S, [], [], Math.log(0.1));

  // L → e₁ Fstar e₂ (paired emission)
  gb.addRule(L, [Fstar], [emit2(1)], 0.0);

  // F* → L S (continue stem or start new)
  gb.addRule(Fstar, [L, S], [], Math.log(0.3));
  // F* → F (switch to unpaired)
  gb.addRule(Fstar, [F], [], Math.log(0.6));
  // F* → ε
  gb.addRule(Fstar, [], [], Math.log(0.1));

  // F → e F (left-emit unpaired)
  gb.addRule(F, [F], [emit1(0)], Math.log(0.8));
  // F → e (terminal unpaired)
  gb.addRule(F, [], [emit1(0)], Math.log(0.2));

  return gb.build(S, 2);
}

/**
 * Gene finder HMM.
 * Models: 0=intergenic, 1=exon, 2=intron.
 */
export function geneFinderGrammar(): Grammar {
  const gb = new GrammarBuilder();
  const IG = gb.addNonterminal('intergenic');
  const E1 = gb.addNonterminal('exon_1');
  const E2 = gb.addNonterminal('exon_2');
  const E3 = gb.addNonterminal('exon_3');
  const I = gb.addNonterminal('intron');

  // Intergenic
  gb.addRule(IG, [IG], [emit1(0)], Math.log(0.95));
  gb.addRule(IG, [E1], [emit1(0)], Math.log(0.04));
  gb.addRule(IG, [], [emit1(0)], Math.log(0.01));

  // Exon phase 1→2→3→1 (codon cycle)
  gb.addRule(E1, [E2], [emit1(1)], Math.log(0.85));
  gb.addRule(E1, [], [emit1(1)], Math.log(0.15));

  gb.addRule(E2, [E3], [emit1(1)], Math.log(0.98));
  gb.addRule(E2, [], [emit1(1)], Math.log(0.02));

  gb.addRule(E3, [E1], [emit1(1)], Math.log(0.80));
  gb.addRule(E3, [I], [emit1(1)], Math.log(0.10));
  gb.addRule(E3, [IG], [emit1(1)], Math.log(0.08));
  gb.addRule(E3, [], [emit1(1)], Math.log(0.02));

  // Intron
  gb.addRule(I, [I], [emit1(2)], Math.log(0.95));
  gb.addRule(I, [E1], [emit1(2)], Math.log(0.04));
  gb.addRule(I, [], [emit1(2)], Math.log(0.01));

  return gb.build(IG, 3);
}

/**
 * Pseudoknot MCFG (fan-out 2).
 * Models: 0=unpaired, 1=paired.
 */
export function pseudoknotGrammar(maxSpan = -1): Grammar {
  const gb = new GrammarBuilder();
  const S = gb.addNonterminal('S');
  const L = gb.addNonterminal('L');
  const PK = gb.addNonterminal('PK', 2, maxSpan);
  const PKinner = gb.addNonterminal('PK*', 2, maxSpan);

  // S → L S
  gb.addRule(S, [L, S], [], Math.log(0.3));
  // S → e S (single-stranded)
  gb.addRule(S, [S], [emit1(0)], Math.log(0.3));
  // S → PK (start pseudoknot — concat fan-out 2 → fan-out 1)
  gb.addRule(S, [PK], [], Math.log(0.1));
  // S → e
  gb.addRule(S, [], [emit1(0)], Math.log(0.2));
  // S → ε
  gb.addRule(S, [], [], Math.log(0.1));

  // L → e₁ S e₂ (paired)
  gb.addRule(L, [S], [emit2(1)], 0.0);

  // PK → PK* (unary)
  gb.addRule(PK, [PKinner], [], 0.0);

  // PK*(ax, by) → PK*(x, y) with composition 'll'
  gb.addRule(PKinner, [PKinner], [emit2(1)], Math.log(0.6), 1);
  // PK*(xa, yb) → PK*(x, y) with composition 'rr'
  gb.addRule(PKinner, [PKinner], [emit2(1)], Math.log(0.2), 2);
  // PK*(x, y) → S(x) S(y) (base case: two fan-out 1 regions)
  gb.addRule(PKinner, [S, S], [], Math.log(0.2));

  return gb.build(S, 2);
}
